_posts/2019-09-30-bert-r/bert_r.Rmd

```diff
@@ -0,0 +1,544 @@
+reticulate::py_config()
+```
+
+```{r eval=F}
+python:         C:/Users/turgut.abdullayev/AppData/Local/Continuum/anaconda3/python.exe
+libpython:      C:/Users/turgut.abdullayev/AppData/Local/Continuum/anaconda3/python37.dll
+pythonhome:     C:\Users\TURGUT~1.ABD\AppData\Local\CONTIN~1\ANACON~1
+version:        3.7.3 (default, Mar 27 2019, 17:13:21) [MSC v.1915 64 bit (AMD64)]
+Architecture:   64bit
+numpy:          C:\Users\TURGUT~1.ABD\AppData\Local\CONTIN~1\ANACON~1\lib\site-packages\numpy
+numpy_version:  1.16.4
+
+NOTE: Python version was forced by use_python function
+```
+
+Luckily for us, a convenient way of importing BERT with Keras was created by Zhao HG. It is called [Keras-bert](https://github.com/CyberZHG/keras-bert). For us, this means that importing that same python library with `reticulate` will allow us to build a popular state-of-the-art model within R.
+
+There are several methods to install keras-bert in Python. 
+
+- in Jupyter Notebook, run:
+
+```{r eval=FALSE}
+!pip install keras-bert
+```
+
+- in Terminal (Linux, Mac OS), run:
+
+```{r eval=FALSE}
+python3 -m pip install keras-bert
+```
+
+- in Anaconda prompt (Windows), run:
+
+```{r eval=FALSE}
+conda install keras-bert
+```
+
+After this procedure, you can check whether keras-bert is installed or not.
+
+```{r eval=F}
+reticulate::py_module_available('keras_bert')
+```
+
+```{r eval=F}
+[1] TRUE
+```
+
+Finally, the TensorFlow version used should be 1.14/1.15. You can check it in the following form:
+
+```{r eval=F}
+tensorflow::tf_version()
+```
+
+```{r eval=F}
+[1] ‘1.14’
+```
+
+In a nutshell:
+
+```{r eval=F}
+pip install keras-bert
+tensorflow::install_tensorflow(version = "1.15")
+```
+
+## What is BERT?
+
+BERT^[[BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/pdf/1810.04805.pdf)] is a pre-trained deep learning model introduced by Google AI Research which has been trained on Wikipedia and BooksCorpus. It has a unique way to understand the structure of a given text. Instead of reading the text from left to right or from right to left, BERT via attention mechanism which is called Transformer encoder^[[Attention Is All You Need](https://arxiv.org/pdf/1706.03762.pdf)] reads the entire word sequences at once. So, it allows to understanding a word based on its surroundings. There are different kind of pre-trained BERT models but the main difference between them is trained parameters. In our case, [BERT](https://storage.googleapis.com/bert_models/2018_10_18/uncased_L-12_H-768_A-12.zip) with 12 encoder layers (Transformer Blocks), 768-hidden hidden units, 12-heads^[Attention — focuses on salient parts of input by taking a weighted average of them. 768 hidden units divided by 12 chunks and each chunk will have 64 output dimensions, afterward, the result from each chunk will be concatenated and forwarded to the next layer], and 110M parameters will be used to create a text classification model.
```

Comment:
```suggestion
BERT^[[BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/pdf/1810.04805.pdf)] is a pre-trained deep learning model introduced by Google AI Research which has been trained on Wikipedia and BooksCorpus. It has a unique way to understand the structure of a given text. Instead of reading the text from left to right or from right to left, BERT, using an attention mechanism which is called Transformer encoder^[[Attention Is All You Need](https://arxiv.org/pdf/1706.03762.pdf)], reads the entire word sequences at once. So, it allows to understanding a word based on its surroundings. There are different kind of pre-trained BERT models but the main difference between them is trained parameters. In our case, [BERT](https://storage.googleapis.com/bert_models/2018_10_18/uncased_L-12_H-768_A-12.zip) with 12 encoder layers (Transformer Blocks), 768-hidden hidden units, 12-heads^[Attention — focuses on salient parts of input by taking a weighted average of them. 768 hidden units divided by 12 chunks and each chunk will have 64 output dimensions, afterward, the result from each chunk will be concatenated and forwarded to the next layer], and 110M parameters will be used to create a text classification model.
```